metabase/metabase
# Feature | Collection types for Metabase Enterprise

This PR adds collection types to Metabase Enterprise Edition. A collection can be either regular or official. Official collections are intended to hold "canonical" items created by the experts in an organization: the data analysts and data scientists, typically.

There is not too much to say about regular collections — they behave the same way collections would do on `master`. On the other hand, official collections and their contents have a few new behaviors:

* they and their contents show up more frequently in search suggestions and results
* official collections and their contents have an official badge in different contexts and pages
* root, personal collections, and personal collections' children can't be official

This is a feature branch, so it contains a set of smaller PRs:

* #17070 
* #17071 
* #17125 
* #17168 
* #17167 
* #17170 

Closes #17072 (epic)
Backend PR: #16191

## How to Test This Feature

First of all, you need to spin up Metabase Enterprise and sign in as an admin. Then, you can follow every step from the feature demo video and ensure everything works the same way

https://user-images.githubusercontent.com/17258145/126975036-b562b3f9-9ae0-4edf-ab32-ed0083c86e73.mov

### Steps for different user flows

<details>
<summary>Creating an official collection</summary>
1. Go to `/collection/root`
2. Click a folder item in the top right ("New collection")
3. The modal form should now have a "Collection type" field at the bottom. "Regular" should be selected by default
4. Click "Official", fill in a collection name, and click "Create"
5. Open the created collection
6. There has to be a yellow ribbon icon next to the collection's name and in the sidebar
</details>

<details>
<summary>Making existing collection official</summary>
1. Open any non-official collection
2. Click a pencil icon in the top right ("Edit collection")
3. Select "Official" collection type and click "Update"
4. There has to be a yellow ribbon icon next to the collection's name
</details>

<details>
<summary>Bulk collection type update</summary>
1. Create a few nested _regular_ collections, let's say A > B > C (B is a child of A, C is a child of B)
2. Open collection A, click the pencil icon in the top right (Edit collection)
3. Click "Official", a checkbox with a "Make children Official too" label should appear
4. Click "Regular" back, the checkbox has to disappear
5. Click "Official" again and click the "Update" button
6. Ensure A, B and C became official (you should see a yellow official badge in the sidebar)
7. Click the pencil icon in the top right (Edit collection)
8. Click "Regular", a checkbox with a "Make children Regular too" label should appear
9. Click the "Update" button
10. Ensure A, B and C became regular again (badge icons have to be replaced with regular blue folder icons)
11. Try repeating the steps above for other levels of nesting (e.g. try updating B and its children, collection A has to keep its original type)
</details>

<details>
<summary>Collection type form field visibility</summary>
1. Visit `/collection/root`
2. Create an official collection
    * click the "new folder" icon in the top right
    * select "Official" at the bottom of a new collection modal, enter a name, click "Create"
    * it should work as in any other PR from #17072
3. Open your personal collection
4. Click the "new folder" icon in the top right. You shouldn't see a "Collection type" form field
6. Enter a name and create a personal collection anyway. Then open it and click the "pencil" icon in the top right. You shouldn't see a "Collection type" form field
10. Sign out and sign in as a normal user
11. Visit `/collection/root`
12. You should see official collections, but when you're trying to create a new collection or edit an existing one, you shouldn't see a "Collection type" field
</details>

<details>
<summary>Search priority</summary>
1. Create an official "Marketing Official" collection
2. Create a regular "Marketing Regular" collection
3. Create any question, call it "Orders" and add it to the "Marketing Official" collection
4. Duplicate the question and save it to the "Marketing Regular" collection
5. Create any dashboard, call it "Orders Dashboard" and add it to the "Marketing Official" collection
6. Duplicate the dashboard and save it to the "Marketing Regular" collection
7. Search for "Marketing", ensure "Marketing Official" is higher than "Marketing Regular"
8. Search for "Orders", ensure that the question and the dashboard from "Marketing Official" are higher than the ones from "Marketing Regular"
</details>

<details>
<summary>Official badge display</summary>

1. Create an official collection
2. Move a question and a dashboard to it
3. Check different places from the demo section below and make sure it looks the same. For regular items (saved to regular collections) it should behave the same way it does on `master`

**Collection sidebar**

![collections-sidebar](https://user-images.githubusercontent.com/17258145/126349394-d1c1058f-998a-41d2-853a-482b49a7eab3.png)

**Question page**

![question-page](https://user-images.githubusercontent.com/17258145/126350003-a91d7729-2ece-4fea-a920-02547769e066.png)

**Dashboard page**

![dashboard-page](https://user-images.githubusercontent.com/17258145/126349451-1fbe5a7e-269c-45fd-9a5a-521899d5dab1.png)

**Search**

![search](https://user-images.githubusercontent.com/17258145/126349741-10c2bb1e-65a1-491b-bb7a-43425f08a7b9.png)

**Saved question picker**

![saved-question-picker](https://user-images.githubusercontent.com/17258145/126349813-93860070-2f1c-4d25-a1de-b198000d9a6a.png)

**Collection picker**

![collection-picker](https://user-images.githubusercontent.com/17258145/126349854-1bc2a84d-6272-4f2e-ad8d-1cd924334165.png)

**Question picker**

![question-picker](https://user-images.githubusercontent.com/17258145/126349888-faa5726e-3180-4be9-aa95-f91e2e9674d0.png)

</details>

<details>
<summary>Displaying official question's badge when it's in regular dashboard</summary>
1. Create a question (the query doesn't really matter) and save it to "Our analytics"
2. Create a dashboard, add the new question to it and save it to "Our analytics"
3. Ensure the question card is displayed as it would `master`
4. Move the question from step 1 to an official collection
5. Open the dashboard from step 2
6. Ensure the question has an official badge next to its name
7. Move the dashboard to an official collection
8. Ensure the question's official badge disappeared
</details>

### Back to OSS

When running Metabase EE, create at least one official collection with at least one question and dashboards in it. Shut down Metabase and spin up Metabase OSS with the same application database. In that way, you would still have a collection with `authority_level: "official"`. The next goal is to make sure collection types don't appear anywhere on the UI. Follow the EE steps and make sure you don't see a "collection type" form field or official "ribbon" icon anywhere. You should still see a collection and its contents, but they should look the same way as if they were regular
